MvvmCross/Base/MvxMainThreadAsyncDispatcher.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MS-PL license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace MvvmCross.Base
+{
+    public abstract class MvxMainThreadAsyncDispatcher : MvxMainThreadDispatcher, IMvxMainThreadAsyncDispatcher
+    {
+        public Task ExecuteOnMainThreadAsync(Action action, bool maskExceptions = true)
+        {
+            var asyncAction = new Func<Task>(() =>
+            {
+                action();
+                return Task.CompletedTask;
+            });
+            return ExecuteOnMainThreadAsync(asyncAction, maskExceptions);
+        }
+
+        public async Task ExecuteOnMainThreadAsync(Func<Task> action, bool maskExceptions = true)
+        {
+            var completion = new TaskCompletionSource<bool>();
+            var syncAction = new Action(async () =>
+            {
+                await action();
+                completion.SetResult(true);
+            });
+            RequestMainThreadAction(syncAction, maskExceptions);
+
+            // If we're already on main thread, then the action will
+            // have already completed at this point, so can just return
+            if (completion.Task.IsCompleted)
```

Comment:
Can't we use ValueTask to do this?

Reply:
`TaskCompletionSource` doesn't return a `ValueTask`? What are you proposing?

Reply:
If we use a `ValueTask` we could return the completed status right away if the Task was finished. Not sure if it makes sense to do here.